### ISSUE TITLE
Memory Inspector show nothing on GB or GBC games

### ISSUE THREAD

**Issue by Altomar79** [RetroAchievements/RALibretro#40]
I tried to load a few different titles and so far i was able to view the memory for one of them. Memory inspector show nothing. The only think that i've notice 
compared to RAVBA is that RALibretro is aware for 8192 RAM Locations for gameboy and 32768 for gameboy color unlike VBA who detect 65536 RAM locations.
Used RALibRetro 1.0.7.180501 with Gambatte core.

![lufia gba](https://user-images.githubusercontent.com/36702297/43815686-98a34516-9ad9-11e8-8098-8e6d38550074.jpg)

**Comment by rzumer** [comment]
The Memory Inspector defaults to displaying the first address in a Note. I assume that most Notes are based on offset memory (on VBA?). Clear the value being watched or compensate for the 0x00c000 offset to view the correct ones.

**Comment by Altomar79** [comment]
Seems like this happen when the displaying note is trying to locate an address outside of the 0x7FFF.
Though GB/GBC have a range up to 0xFFFF

**Comment by Altomar79** [comment]
Also RALibretro commit hash is 75845752317df71187b821929f2f848980b34b00 in case this help.
Its the latest build we have in our download page.

**Comment by Jamiras** [comment]
I took a look at this, and while I understand what's happening, I'm not sure how to fix it. Someone more familiar with the libretro cores will have to address this.

The basic problem is that the GB/GBC cores only map the System RAM to the integration DLL as $0000-$7FFF. 

This is wrong. RAVBA-M actually registers the full address space for GB/GBC, so the System RAM should be mapped from $C000-$DFFF and echoed at $E000-$FDFF. I'm not sure why `_core.GetMemorySize(RETRO_MEMORY_SYSTEM_RAM)` is returning 32K, I think it should only be 8K. 

Additionally, the Save RAM is completely ignored by RALibRetro and needs to be mapped to $A000-$BFFF. And I have no idea what needs to be mapped from $0000-$9FFF.

The solution to this seems to be to replace the mapping with a raw read for GB/GBC:
```
      RA_InstallMemoryBank(0, (void*)::memoryReadRaw, (void*)::memoryWriteRaw, 0x10000);
```
```
unsigned char Application::memoryReadRaw(unsigned addr)
{
  const struct retro_memory_map* mmap = _core.getMemoryMap();
  const retro_memory_descriptor *desc = mmap->descriptors;
  const retro_memory_descriptor *end = desc + mmap->num_descriptors;
  while (desc < end)
  {
    if (desc->start <= addr && desc->start + desc->len > addr)
    {
      unsigned char* memory = (unsigned char*)desc->ptr + desc->offset;
      return memory[addr - desc->start];
    }

    ++desc;
  }

  return 0;
}
```
This seems to mostly work with the Gambatte core, but reads between $E000 and $FDFF fail because the echo ram isn't registered in the mmap.

The mGBA core doesn't provide a mmap, so fails miserably (does it only support GBA?).

**Comment by meleu** [comment]
> The mGBA core doesn't provide a mmap, so fails miserably (does it only support GBA?).

Yes.

https://docs.libretro.com/guides/retroachievements/#game-boy-game-boy-color

**Comment by meleu** [comment]
solved in eee25ffad74ab7225d1a5143936f50a0c759c733
